### PR TITLE
[gen_l10n] Absolute project path parameter

### PR DESCRIPTION
**Pull request by shihaohong**
## Description

Creates an option for the gen_l10n tool to define the absolute path to the Flutter project that needs its localizations generated. This PR helps support https://github.com/flutter/flutter/pull/62395/, which requires specifying the absolute path to the Flutter project.

## Tests

I added the following tests:
- Test to make sure that the absolute path option is used
- Test to ensure that an error is thrown if the absolute path specified does not exist

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by HansMuller** [comment]
Looks like the default value of project-dir needs some test updates

File: packages/flutter_tools/test/general.shard/build_system/targets/localizations_test.dart [hunk 1]
```diff
@@ -24,6 +24,7 @@ void main() {
           '--disable-dart-dev',
           'dev/tools/localization/bin/gen_l10n.dart',
           '--gen-inputs-and-outputs-list=/',
+          "--project-dir=MemoryDirectory: '/path/to/flutter_project'",
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/build_system/targets/localizations_test.dart:27]: this seems wrong, it should just have the path contents

Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/build_system/targets/localizations_test.dart:27]: possibly an issue with projectDir

File: packages/flutter_tools/lib/src/build_system/targets/localizations.dart [hunk 2]
```diff
@@ -36,11 +36,14 @@ Future<void> generateLocalizations({
     'bin',
     'gen_l10n.dart',
   );
+  final String projectDirPath = projectDir.path.replaceAll("'", '');
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/build_system/targets/localizations.dart:39]: You shouldn't need to manipulate the path string like this - what problem is this solving?

Comment (shihaohong) [reply packages/flutter_tools/lib/src/build_system/targets/localizations.dart:39]: This was a mistake I made. The test failed for some other reason and I couldn't understand why. I'll remove this line in a subsequent commit. Thanks for the catch

**Comment by shihaohong** [comment]
This PR should now be ready for review

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**